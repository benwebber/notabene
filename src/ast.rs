//! A simplified Markdown document AST.
//!
//! The AST is a sequence of [`Block`] elements, associated with spans in the original document.
mod block;
mod parser;

pub(crate) use block::*;
pub(crate) use parser::Parser;
