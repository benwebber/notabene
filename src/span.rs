//! Work with spans within the source document.
use std::ops::{Deref, Range};

use serde::{Deserialize, Serialize};

mod iter;
mod locator;
mod ranged;
mod unist;

pub(crate) use iter::SpanIterator;
pub use locator::Locator;
pub(crate) use ranged::Ranged;
pub use unist::{Point, Position};

/// A span within the source document.
///
/// Use [`Locator`] to convert a `Span` to a [`Position`].
#[derive(
    Copy, Clone, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize,
)]
pub struct Span {
    /// The start offset of the span.
    pub start: usize,
    /// The end offset of the span.
    pub end: usize,
}

#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Serialize)]
pub(crate) struct Spanned<T> {
    pub span: Span,
    pub value: T,
}

impl Ranged<usize> for Span {
    fn range(&self) -> Range<usize> {
        (*self).into()
    }
}

impl<'a> Deref for Spanned<&'a str> {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        self.value
    }
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    /// Offset a span by `offset`.
    pub fn offset(&self, offset: usize) -> Self {
        Self {
            start: offset + self.start,
            end: offset + self.end,
        }
    }
}

impl From<Range<usize>> for Span {
    fn from(range: Range<usize>) -> Self {
        Self {
            start: range.start,
            end: range.end,
        }
    }
}

impl From<Span> for Range<usize> {
    fn from(span: Span) -> Self {
        span.start..span.end
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_offset() {
        let span = Span::new(0, 10);
        assert_eq!(span.offset(5), Span::new(5, 15));
    }

    #[test]
    fn test_range() {
        let span = Span::new(0, 10);
        assert_eq!(span.range(), span.into())
    }
}
