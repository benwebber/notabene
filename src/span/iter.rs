use super::Span;

pub struct SpanIterator<'a> {
    text: &'a str,
    cursor: usize,
}

impl<'a> SpanIterator<'a> {
    pub fn new(text: &'a str) -> Self {
        Self { text, cursor: 0 }
    }
}

impl<'a> Iterator for SpanIterator<'a> {
    type Item = Span;

    fn next(&mut self) -> Option<Self::Item> {
        let start = self.text[self.cursor..]
            .char_indices()
            .find(|(_, c)| !c.is_whitespace())
            .map(|(i, _)| self.cursor + i);
        let Some(start) = start else {
            // End of string.
            self.cursor = self.text.len();
            return None;
        };
        let end = self.text[start..]
            .char_indices()
            .find(|(_, c)| c.is_whitespace())
            .map(|(i, _)| start + i)
            .unwrap_or(self.text.len());
        self.cursor = end;
        Some(Span::new(start, end))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_span_iterator() {
        let s = "  foo  foobar  quux  ";
        let spans: Vec<Span> = SpanIterator::new(s).collect();
        assert_eq!(
            spans,
            vec![Span::new(2, 5), Span::new(7, 13), Span::new(15, 19)]
        );
    }
}
