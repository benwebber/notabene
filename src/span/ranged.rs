pub trait Ranged<T> {
    fn range(&self) -> std::ops::Range<T>;
}

impl Ranged<usize> for std::ops::Range<usize> {
    fn range(&self) -> Self {
        self.clone()
    }
}
