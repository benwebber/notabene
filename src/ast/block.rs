//! AST block elements.
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

use crate::span::Span;

/// A block element.
#[derive(Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
pub enum Block {
    Heading(Heading),
    Paragraph(Literal),
    List(List),
}

/// An inline element.
#[derive(Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
pub enum Inline {
    Link(Link),
    Literal(Literal),
}

/// A section heading.
#[derive(Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
pub struct Heading {
    pub span: Span,
    pub level: usize,
    pub inlines: Vec<Inline>,
}

/// A link.
#[derive(Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
pub struct Link {
    pub span: Span,
    pub content: Literal,
    // TODO: Figure out how to obtain span for this.
    pub target: String,
}

/// Literal markup.
#[derive(Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
pub struct Literal {
    pub span: Span,
}

/// An unordered list.
#[derive(Debug, PartialEq)]
#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
pub struct List {
    pub span: Span,
    pub items: Vec<Literal>,
}
