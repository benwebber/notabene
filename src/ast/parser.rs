//! Parse a Markdown document as an iterator of block elements.
use pulldown_cmark as md;

use super::*;
use crate::span::Span;

/// A [pulldown-cmark](https://crates.io/crates/pulldown-cmark) parser that includes text offset
/// information (for spans) and merges consecutive text nodes into one.
type MdParser<'a> = md::utils::TextMergeWithOffset<'a, md::OffsetIter<'a>>;

pub struct Parser<'a> {
    inner: MdParser<'a>,
}

impl<'a> Parser<'a> {
    pub fn new(s: &'a str) -> Self {
        let parser = md::Parser::new(s).into_offset_iter();
        let inner = md::utils::TextMergeWithOffset::new(parser);
        Self { inner }
    }
}

impl Literal {
    fn new<S: Into<Span>>(span: S) -> Self {
        Self { span: span.into() }
    }
}

impl Iterator for Parser<'_> {
    type Item = Block;

    fn next(&mut self) -> Option<Self::Item> {
        while let Some((event, range)) = self.inner.next() {
            match event {
                md::Event::Start(md::Tag::Heading { level, .. }) => {
                    let inlines = parse_inlines(&mut self.inner, md::TagEnd::Heading(level));
                    let level = match level {
                        md::HeadingLevel::H1 => 1,
                        md::HeadingLevel::H2 => 2,
                        md::HeadingLevel::H3 => 3,
                        md::HeadingLevel::H4 => 4,
                        md::HeadingLevel::H5 => 5,
                        md::HeadingLevel::H6 => 6,
                    };
                    // Remove trailing newline.
                    let span = Span::new(range.start, range.end - 1);
                    let heading = Heading {
                        span,
                        level,
                        inlines,
                    };
                    return Some(Block::Heading(heading));
                }
                md::Event::Start(md::Tag::Paragraph) => {
                    let lit = Literal::new(range);
                    return Some(Block::Paragraph(lit));
                }
                md::Event::Start(md::Tag::List(None)) => {
                    let items = parse_items(&mut self.inner);
                    let list = List {
                        span: range.into(),
                        items,
                    };
                    return Some(Block::List(list));
                }
                _ => {}
            }
        }
        None
    }
}

fn parse_inlines(parser: &mut MdParser, until: md::TagEnd) -> Vec<Inline> {
    let mut inlines = vec![];
    while let Some((event, span)) = parser.next() {
        match event {
            md::Event::Start(md::Tag::Link { dest_url, .. }) => {
                let content_span = read_span_until(parser, md::TagEnd::Link);
                let content = Literal::new(content_span);
                let target: String = dest_url.into();
                let link = Link {
                    span: span.into(),
                    content,
                    target,
                };
                inlines.push(Inline::Link(link))
            }
            md::Event::Text(_) => {
                let span = Span::from(span);
                inlines.push(Inline::Literal(Literal::new(span)));
            }
            md::Event::End(tag) if tag == until => break,
            _ => {}
        }
    }
    inlines
}

fn parse_items(parser: &mut MdParser) -> Vec<Literal> {
    let mut items = vec![];
    while let Some((event, _)) = parser.next() {
        match event {
            md::Event::Start(md::Tag::Item) => {
                let item_span = read_span_until(parser, md::TagEnd::Item);
                items.push(Literal::new(item_span));
            }
            md::Event::End(md::TagEnd::List(false)) => break,
            _ => {}
        }
    }
    items
}

fn read_span_until(parser: &mut MdParser, until: md::TagEnd) -> Span {
    let mut span = Span::default();
    for (event, span_) in parser.by_ref() {
        match event {
            md::Event::End(tag) if tag == until => {
                break;
            }
            _ => {
                if span.start == 0 {
                    span.start = span_.start
                }
                span.end = span_.end
            }
        }
    }
    span
}

#[cfg(test)]
mod tests {

    use super::*;

    macro_rules! snapshot {
        ($s:literal) => {
            let blocks: Vec<Block> = Parser::new(&$s.trim()).collect();
            insta::assert_yaml_snapshot!(blocks);
        };
    }

    #[test]
    fn test_parse_h1() {
        snapshot!("# Changelog");
        snapshot!(
            "
Changelog
=========
        "
        );
    }

    #[test]
    fn test_parse_h2() {
        snapshot!(
            "
## [Unreleased]

[Unreleased]: https://example.org/
        "
        );
        snapshot!(
            "
[Unreleased]
------------

[Unreleased]: https://example.org/
        "
        );
        snapshot!("## [Unreleased](https://example.org/)");
        snapshot!(
            "
[Unreleased](https://example.org/)
----------------------------------
"
        );
        snapshot!(
            "
## [1.6.8] - 2038-01-19 [YANKED]

[1.6.8]: https://example.org/
"
        );
        snapshot!(
            "
[1.6.8] - 2038-01-19 [YANKED]
-----------------------------

[1.6.8]: https://example.org/
"
        );
        snapshot!("## [1.6.8](https://example.org/) - 2038-01-19 [YANKED]");
        snapshot!(
            "
[1.6.8](https://example.org/) - 2038-01-19 [YANKED]
---------------------------------------------------
"
        );
    }

    #[test]
    fn test_parse_h3() {
        snapshot!("### Added");
    }

    #[test]
    fn test_parse_paragraph() {
        snapshot!(
            "
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
        ");
    }

    #[test]
    fn test_parse_list() {
        snapshot!(
            "
* Lorem ipsum dolor sit amet
* Consectetur adipiscing elit
* Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua
        "
        );
    }
}
