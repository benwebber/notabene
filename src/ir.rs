//! Intermediate representation (IR) of a changelog.
//!
//! Parsing a changelog returns an [`ir::Changelog`](crate::ir::Changelog).
//! This model differs from the [`Changelog`](crate::Changelog) in two significant ways:
//!
//!   * Elements include span information.
//!   * It allows multiple titles and unreleased sections.
//!
//! Preserving span information allows lint checks to include the context around diagnostics.
use crate::span::Span;

use serde::Serialize;

#[derive(Debug, Default, Eq, Hash, PartialEq, Serialize)]
pub(crate) struct Spanned<T> {
    pub span: Span,
    pub value: T,
}

#[derive(Debug, Default, Serialize)]
pub(crate) struct Changelog<'a> {
    pub sections: Vec<Section<'a>>,
    pub broken_links: Vec<Span>,
}

#[derive(Debug, Serialize)]
pub enum Section<'a> {
    Title(Spanned<&'a str>),
    Unreleased(Unreleased<'a>),
    Release(Release<'a>),
    Invalid(InvalidSection),
    // TODO: If Title held something like Vec<Inline>, it would be possible
    // to get rid of InvalidTitle and validate Title in the linter.
    InvalidTitle(InvalidTitle),
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Unreleased<'a> {
    pub heading_span: Span,
    // TODO: Get span.
    pub url: Option<String>,
    pub changes: Vec<Changes<'a>>,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Release<'a> {
    pub heading_span: Span,
    pub version: Spanned<&'a str>,
    // TODO: Get span.
    pub url: Option<String>,
    pub date: Option<Spanned<&'a str>>,
    pub yanked: Option<Spanned<&'a str>>,
    pub changes: Vec<Changes<'a>>,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Changes<'a> {
    pub heading_span: Span,
    pub kind: Spanned<&'a str>,
    pub changes: Vec<Spanned<&'a str>>,
}

impl<T> Spanned<T> {
    pub(crate) fn new(span: Span, value: T) -> Self {
        Self { value, span }
    }

    pub(crate) fn into_inner(self) -> T {
        self.value
    }
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct InvalidSection {
    pub heading_span: Span,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct InvalidTitle {
    pub heading_span: Span,
}
