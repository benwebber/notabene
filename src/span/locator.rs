use std::cmp::Ordering;

use super::{Point, Position, Ranged, Span};

/// Locates spans in the source document.
pub struct Locator<'a> {
    source: &'a str,
    line_spans: Vec<Span>,
}

impl<'a> Locator<'a> {
    pub fn new(source: &'a str) -> Self {
        let mut line_spans = Vec::new();
        let mut start = 0;
        for (end, _) in source.match_indices('\n') {
            line_spans.push(Span::new(start, end));
            start = end + 1;
        }
        if start <= source.len() {
            // No trailing newline.
            line_spans.push(Span::new(start, source.len()))
        }
        Self { source, line_spans }
    }

    /// Return the unist Point for the given offset.
    fn point(&self, offset: usize) -> Point {
        if self.line_spans.is_empty() {
            return Point::default();
        }
        let cmp = |span: &Span| {
            if offset < span.start {
                Ordering::Greater
            } else if offset > span.end {
                Ordering::Less
            } else {
                Ordering::Equal
            }
        };
        let i = match self.line_spans.binary_search_by(cmp) {
            Ok(i) => i,
            Err(i) => i
                .saturating_sub(1)
                .min(self.line_spans.len().saturating_sub(1)),
        };
        Point::new(
            i + 1,
            offset.saturating_sub(self.line_spans[i].start) + 1,
            offset,
        )
    }

    /// Return the unist Position for the given span.
    pub fn position<R: Ranged<usize>>(&self, ranged: &R) -> Position {
        let range = ranged.range();
        Position::new(self.point(range.start), self.point(range.end))
    }

    /// Return the contents of line number `line`.
    pub(crate) fn line(&self, line: usize) -> &'a str {
        &self.source[self.line_spans[line.saturating_sub(1)].range()]
    }

    /// Return the number of lines in the source.
    pub(crate) fn lines(&self) -> usize {
        self.line_spans.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::span::SpanIterator;

    #[test]
    fn test_position() {
        let s = "foo

bar

quux";
        let locator = Locator::new(s);
        let spans: Vec<Span> = SpanIterator::new(s).collect();
        assert_eq!(
            locator.position(&spans[0]),
            Position::new(Point::new(1, 1, 0), Point::new(1, 4, 3)),
        );
        assert_eq!(
            locator.position(&spans[1]),
            Position::new(Point::new(3, 1, 5), Point::new(3, 4, 8)),
        );
        assert_eq!(
            locator.position(&spans[2]),
            Position::new(Point::new(5, 1, 10), Point::new(5, 5, 14)),
        );
    }

    #[test]
    fn test_line() {
        let s = "foo bar baz\nfoobar foobaz";
        let locator = Locator::new(s);
        assert_eq!(locator.line(2), "foobar foobaz");
    }
}
