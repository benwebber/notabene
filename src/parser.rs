//! Parse a changelog as its [intermediate representation](crate::ir::Changelog).
use crate::ast::{self, *};
use crate::ir::*;
use crate::span::{Span, SpanIterator};
use std::cell::RefCell;
use std::iter::Peekable;

use pulldown_cmark as md;

/// Parse a changelog.
///
/// Parsing never fails.
pub fn parse(s: &str) -> Changelog {
    let mut changelog = Changelog::default();
    let broken_links = RefCell::new(Vec::new());
    let callback = |link: md::BrokenLink| {
        broken_links.borrow_mut().push(link.span.into());
        None
    };
    let parser = md::Parser::new_with_broken_link_callback(s, md::Options::empty(), Some(callback));
    let parser = md::utils::TextMergeWithOffset::new(parser.into_offset_iter());
    let mut blocks = ast::Parser::new(Box::new(parser)).peekable();
    while let Some(block) = blocks.next() {
        match block {
            Block::Heading(heading @ Heading { level: 1, .. }) => {
                let section = match get_heading_span(&heading) {
                    Some(span) => {
                        let spanned = Spanned::new(span, s[span.range()].to_string());
                        Section::Title(spanned)
                    }
                    None => Section::InvalidTitle(InvalidTitle {
                        heading_span: heading.span,
                    }),
                };
                changelog.sections.push(section);
            }
            Block::Heading(heading @ Heading { level: 2, .. }) => {
                let section = parse_section(s, &heading, &mut blocks);
                if let Some(sec) = section {
                    changelog.sections.push(sec);
                }
            }
            _ => {}
        }
    }
    // `blocks` still holds a reference to `callback` through the parser.
    drop(blocks);
    changelog.broken_links = broken_links.into_inner();
    changelog
}

fn parse_section<'a>(
    s: &str,
    heading: &Heading,
    blocks: &mut Peekable<ast::Parser<'a>>,
) -> Option<Section> {
    match heading.inlines.as_slice() {
        // Unreleased
        [Inline::Link(l)] if &s[l.content.span.range()] == "Unreleased" => {
            let changes = parse_changes(s, blocks);
            Some(Section::Unreleased(Unreleased {
                heading_span: heading.span,
                url: Some(l.target.clone()),
                changes,
            }))
        }
        // Release
        [Inline::Link(l), Inline::Literal(t)] => {
            let mut release = Release {
                heading_span: heading.span,
                version: Spanned::new(l.content.span, s[l.content.span.range()].to_string()),
                url: Some(l.target.clone()),
                ..Default::default()
            };
            let mut spans = SpanIterator::new(&s[t.span.range()]);
            // Skip hyphen.
            spans.next();
            if let Some(span) = spans.next().map(|s| s.offset(t.span.start)) {
                let date = &s[span.range()];
                release.date = Some(Spanned::new(span, date.to_string()))
            }
            if let Some(span) = spans.next().map(|s| s.offset(t.span.start)) {
                let yanked = &s[span.range()];
                release.yanked = Some(Spanned::new(span, yanked.to_string()));
            }
            let changes = parse_changes(s, blocks);
            release.changes = changes;
            Some(Section::Release(release))
        }
        _ => Some(Section::Invalid(InvalidSection {
            heading_span: heading.span,
        })),
    }
}

fn parse_changes<'a>(s: &str, blocks: &mut Peekable<ast::Parser<'a>>) -> Vec<Changes> {
    let mut sections: Vec<(Span, Spanned<String>, Vec<Spanned<String>>)> = Vec::new();
    let mut current_kind: Option<String> = None;
    let mut current_changes: Vec<Spanned<String>> = Vec::new();
    let mut current_heading_span: Span = Span::default();

    while let Some(block) = blocks.peek() {
        match block {
            Block::Heading(heading @ Heading { level: 3, .. }) => {
                if let Some(kind) = current_kind.take() {
                    // TODO: more accurate span for kind
                    sections.push((
                        current_heading_span,
                        Spanned::new(current_heading_span, kind),
                        std::mem::take(&mut current_changes),
                    ));
                }
                current_kind = get_heading_text(s, heading).map(|s| s.to_string());
                current_heading_span = heading.span;
                blocks.next();
            }
            Block::List(l) => {
                if current_kind.is_some() {
                    current_changes.extend(
                        l.items
                            .iter()
                            .map(|i| Spanned::new(i.span, s[i.span.range()].to_string())),
                    );
                }
                blocks.next();
            }
            _ => break,
        }
    }

    if let Some(kind) = current_kind.take() {
        sections.push((
            current_heading_span,
            Spanned::new(current_heading_span, kind),
            current_changes,
        ));
    }

    sections
        .into_iter()
        .map(|(heading_span, kind, spanned)| Changes {
            heading_span,
            kind,
            changes: spanned,
        })
        .collect()
}

fn get_heading_span(heading: &Heading) -> Option<Span> {
    match heading.inlines.as_slice() {
        [Inline::Literal(Literal { span, .. })] => Some(*span),
        _ => None,
    }
}

fn get_heading_text<'a>(s: &'a str, heading: &Heading) -> Option<&'a str> {
    match heading.inlines.as_slice() {
        [Inline::Literal(Literal { span, .. })] => Some(&s[span.range()]),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    #[test]
    fn test_parse() {
        let source = "
# Title 1
# Title 2

## [Unreleased][]

### Removed

* Remove foo

## [1.0.0] - 2025-01-01

### Added

* Add foo
* Add bar

## [0.1.0] - 2024-01-01

### Added

* Add baz
* Add quux

## [broken] - 2024-01-01

[Unreleased]: https://example.org/unreleased
[1.0.0]: https://example.org/release/1.0.0
[0.1.0]: https://example.org/release/0.1.0
        ";
        let changelog = parse(source);
        assert_yaml_snapshot!(changelog);
    }
}
