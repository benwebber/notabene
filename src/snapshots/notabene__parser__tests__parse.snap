---
source: src/parser.rs
expression: changelog
---
source: "\n# Title 1\n# Title 2\n\n## [Unreleased][]\n\n### Removed\n\n* Remove foo\n\n## [1.0.0] - 2025-01-01\n\n### Added\n\n* Add foo\n* Add bar\n\n## [0.1.0] - 2024-01-01\n\n### Added\n\n* Add baz\n* Add quux\n\n## [broken] - 2024-01-01\n\n[Unreleased]: https://example.org/unreleased\n[1.0.0]: https://example.org/release/1.0.0\n[0.1.0]: https://example.org/release/0.1.0\n        "
title:
  span:
    start: 3
    end: 10
  value: Title 1
unreleased:
  heading_span:
    start: 22
    end: 39
  url: "https://example.org/unreleased"
  changes:
    - heading_span:
        start: 41
        end: 52
      kind:
        span:
          start: 41
          end: 52
        value: Removed
      items:
        - span:
            start: 56
            end: 66
          value: Remove foo
releases:
  - heading_span:
      start: 68
      end: 91
    version:
      span:
        start: 72
        end: 77
      value: 1.0.0
    url: "https://example.org/release/1.0.0"
    date:
      span:
        start: 81
        end: 91
      value: 2025-01-01
    yanked: ~
    changes:
      - heading_span:
          start: 93
          end: 102
        kind:
          span:
            start: 93
            end: 102
          value: Added
        items:
          - span:
              start: 106
              end: 113
            value: Add foo
          - span:
              start: 116
              end: 123
            value: Add bar
  - heading_span:
      start: 125
      end: 148
    version:
      span:
        start: 129
        end: 134
      value: 0.1.0
    url: "https://example.org/release/0.1.0"
    date:
      span:
        start: 138
        end: 148
      value: 2024-01-01
    yanked: ~
    changes:
      - heading_span:
          start: 150
          end: 159
        kind:
          span:
            start: 150
            end: 159
          value: Added
        items:
          - span:
              start: 163
              end: 170
            value: Add baz
          - span:
              start: 173
              end: 181
            value: Add quux
invalid_spans:
  - DuplicateTitle:
      start: 13
      end: 20
  - InvalidSectionHeading:
      start: 183
      end: 207
  - UndefinedLinkReference:
      start: 186
      end: 194
